File: codes/eunjung/784_lettercasepermutation_1.js
function letterCasePermutation(value){

    let allCombination = [];

    function backTrack(path, index){

        if(index === value.length){
            allCombination.push(path);
            return;
        }

        let singleString = value[index];

        if(isNaN(singleString)){// if it is a character
            backTrack(path + singleString.toLowerCase(), index + 1);
            backTrack(path + singleString.toUpperCase(), index + 1);
        }else{//if it is a number
            backTrack(path + singleString, index + 1);
        }
    }

    backTrack('', 0);

    return allCombination;
}

File: codes/naeun/784_lettercasepermutation_1.java
class Solution {
    public List<String> letterCasePermutation(String s) {
        List<String> result = new ArrayList<>();
        backtrack(s.toCharArray(), 0, result);
        return result;
    }

    private void backtrack(char[] s, int index, List<String> result) {
        if (index == s.length) {
            result.add(new String(s));
            return;
        }

        if (Character.isLetter(s[index])) {
            // lowercase
            s[index] = Character.toLowerCase(s[index]);
            backtrack(s, index + 1, result);
            // uppercase
            s[index] = Character.toUpperCase(s[index]);
            backtrack(s, index + 1, result);
        } else {
            backtrack(s, index + 1, result);
        }
    }
}